### PR TITLE
Move internal diophantine solvers into classes

### PR DESCRIPTION
**Pull request by ethankward**
#### References to other Issues or PRs
#18493

#### Brief description of what is fixed or changed
Move all internal solvers (`_diop_linear`, `_diop_quadratic`, `_diop_ternary_quadratic`, `_diop_ternary_quadratic_normal`, `_diop_general_pythagorean`, `_diop_general_sum_of_squares`, `_diop_general_sum_of_even_powers`) into a new `solve` method in `DiophantineEquationType`.

Also added `diop_solve` to `__init__.py`.

Added a new parameter `n_parameters` to `DiophantineSolutionSet` to explicitly control the number of parameters.

#### Other comments
`_diop_ternary_quadratic` and `_diop_ternary_quadratic_normal` have been retained as placeholders for now as they are used in other functions.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/solvers/diophantine/diophantine.py [hunk 1]
```diff
@@ -190,27 +188,391 @@ def matches(self):
...
 
 class Linear(DiophantineEquationType):
     name = 'linear'
 
     def matches(self):
         return self.total_degree == 1
 
+    def solve(self, params=None, limit=None):
+        """
+        Solves diophantine equations of the form:
```
Comment (oscarbenjamin) [inline sympy/solvers/diophantine/diophantine.py]: Would this make more sense as a class-level docstring?

Comment (ethankward) [reply sympy/solvers/diophantine/diophantine.py]: Yes. I will merge this with the docstring that is currently in the `diop_linear` function.

**Comment by oscarbenjamin** [comment]
Are you done with this?

**Comment by ethankward** [comment]
No, I'm going to make documentation changes as you suggested.

**Comment by ethankward** [comment]
OK, I had to make some more changes as I found some bugs. This is ready for review again.

**Comment by oscarbenjamin** [comment]
Do the new docstrings need to be added to the rst docs somewhere?

**Comment by ethankward** [comment]
I would like to avoid modifying the .rst files until this is in a final state, since various things will probably be deprecated.

File: sympy/solvers/diophantine/diophantine.py [hunk 2]
```diff
@@ -98,19 +99,34 @@ class DiophantineSolutionSet(set):
...
+                self.n_parameters = len(parameters)
+
         if parameters is None:
-            self.parameters = symbols('%s1:%i' % ('t', len(self.symbols) + 1), integer=True)
+            self.parameters = symbols('%s1:%i' % ('t', self.n_parameters + 1), integer=True)
         else:
             self.parameters = tuple(parameters)
 
+            if n_parameters is not None:
+                if len(self.parameters) < self.n_parameters:
```
Comment (oscarbenjamin) [inline sympy/solvers/diophantine/diophantine.py]: All the logic above with `n_parameters` and `self.n_parameters` is confusing. Is there a need for `n_parameters` to be anything other than len(self.parameters)

Comment (ethankward) [reply sympy/solvers/diophantine/diophantine.py]: I'll see if I can simplify this. If `parameters` is an optional argument then it's necessary to be able to control its length.

Comment (ethankward) [reply sympy/solvers/diophantine/diophantine.py]: Ok, I have removed `n_parameters` from `DiophantineSolutionSet` and made `parameters` a required argument. Subsequently, I added a `get_parameters` method to the `DiophantineEquationType` that can be overridden in the individual solvers. This greatly simplifies the logic of dealing with the interaction of `n_parameters` / `parameters`.

**Comment by ethankward** [comment]
This PR needs more work but I'm leaving for several days- I will return to this next week.

File: sympy/solvers/diophantine/diophantine.py (hunk 2 above)
Comment (oscarbenjamin) [reply sympy/solvers/diophantine/diophantine.py]: That seems better

**Comment by ethankward** [comment]
I moved `parameters` and `n_parameters` to class properties (so that parameters would not be recreated every time they are accessed). Also moved pre-solution checking to its own method.

**Comment by oscarbenjamin** [comment]
Should the classes here be added to the rst docs somewhere so that the docstrings show up?

**Comment by oscarbenjamin** [comment]
This looks good. Are you done with it?

**Comment by oscarbenjamin** [comment]
Is this finished?

**Comment by czgdp1807** [comment]
@ethankward @oscarbenjamin Can we merge this?

File: sympy/solvers/diophantine/diophantine.py [hunk 3]
```diff
@@ -98,18 +97,17 @@ class DiophantineSolutionSet(set):
         {(t**2, t + 2)}
     """
 
-    def __init__(self, symbols_seq, parameters=None):
+    def __init__(self, symbols_seq, parameters):
```
Comment (czgdp1807) [inline sympy/solvers/diophantine/diophantine.py:100]: IMO, this is a public API change. Do we need a deprecation warning for this?

Comment (czgdp1807) [reply sympy/solvers/diophantine/diophantine.py:100]: @oscarbenjamin What do you say about this change and the overall PR?

Comment (oscarbenjamin) [reply sympy/solvers/diophantine/diophantine.py:100]: I think the PR looks good. We should resolve the merge conflicts and get it in.

**Comment by oscarbenjamin** [comment]
Thanks @ethankward for this. I think probably you were intending to do more with this but what is here so far is still good work and it's best to get it in so that others can contribute further.

I'll merge this now.

**Pull request merged**